### PR TITLE
[READY/TEST] More HUD Implants, less Thermal Vision

### PR DESCRIPTION
**Pull request by WanderingFox95**
## About The Pull Request
I... actually meant to post this as a PR to my fork to first ask others for an opinion on it, but now I accidentally posted it already.
So... yeah, here's a PR.

Since the dawn of time, people complain about thermal vision and how overabundant it is. I wanted to just add science goggle implants and diagnostic hud implants, but then I also added Meson Vision Eyes and that gave me an idea. Why not knock everything up one level? Mesons for combat implants, illegal tech implants get thermals and alien biotech gets X-ray eyes. And because only removing it from implants wasn't satisfying enough, I made meson vision a mutation and replaced thermals with that too!

Edit: Just so this is clear: I know the 'Focus' has become the Thermals rebalance, but actually, it started as an Implant PR. After I added Meson eyes, I couldn't stop thinking that it could be used to balance things for the better. Depending on how it's decided, the rebalance will happen or will be removed, but the implants and mesons can absolutely stay either way. It's a matter of balance and what the discussion results in.

## Why It's Good For The Game

Why IS there no diagnostics HUD implant? Why ARE there no MESON eyes? I wanted to stay giving all the goggle/hud types implants. While doing this, I felt compelled to use them to balance the subject of many arguments and much salt in the past: Thermal vision. The traitor uplink contains thermals. This means, unlocking them with illegal tech makes sense. Since they were also in genetics, I abruptly made a meson mutation and replaced thermals with that too. It makes sense, it's good for balance, and you're encouraged to get more surgeries done on you. I believe it would only add to gameplay.

## Changelog
:cl:
add: Science and Diagnostics HUD implants. Also cybernetic tongues are back.
add: Meson Eyes, Meson Vision mutation, Toggleable Nightvision Mutation.
add: The CE now starts with a Meson Eye Autosurgeon, the RD with a diagnostics HUD implant Autosurgeon
tweak: The eye-cybernetics were all bumped up by one research tier. I mean, thermals ARE available in the uplink, it's only fair. x-ray eyes are now part of alien technology research.
tweak: The mutation recipe for thermal vision (Nearsighted+Stimmed) now produces night vision instead. (Thermals can still be granted with admin-spawned injectors)
balance: Overall, Thermals are now more difficult to get but they are NOT removed entirely!
/:cl:

### CODE REVIEW

**Comment by Solaris-Shade** [comment]
Thinking about the wheel of salt and all of the funny things it says.....There is always a nugget of truth in that piles of salt. And Thermals have always been MOUNTAIN RANGES of salty discussions.

That being said.... I feel that this is a good step in the right direction.

**Comment by ghost** [comment]
This is a pretty good compromise. Innate mesons mean you can still see the layout of the station while still giving antagonists the ability to stealth, albeit slightly harder if it involves false walls due to mesons. Locking thermals behind illegals also means that crew can't simply hand them out willy nilly without a traitor making a huge mistake, which again is very nice.

All in all I'm a big fan of this

**Comment by silicons** [comment]
mmmmm

**Comment by silicons** [comment]
tentative approval + upvote from me

**Comment by BlueWildrose** [comment]
I'm unsure if genetics meson vision warrants 40 instability, genetics-wise - nor does it need to be locked behind a mutation of near-sighted+stimmed, compared to thermal vision. It doesn't give away the location of antagonists at all aside from the aforementioned false walls, and meson scanners on their own are very much prevalent even as just possible maintenance loot. You arguably might as well just wear the meson scanners/engineering goggles/NVG mesons. It's too expensive for the benefit it gives under the current setup. It's nowhere near the same as thermal vision, which was made expensive intentionally due to how OP it was.

Also, as for the meson eyes - are these eyes subject to EMPs? _Wouldn't that be an extremely bad idea for the chief engineer if they, for example wanted to work with a singularity, which emits EMPs often_?

**Comment by ghost** [comment]
> 
> 
> I'm unsure if genetics meson vision warrants 40 instability, genetics-wise - nor does it need to be locked behind a mutation of near-sighted+stimmed, compared to thermal vision. It doesn't give away the location of antagonists at all aside from the aforementioned false walls, and meson scanners on their own are very much prevalent even as just possible maintenance loot. You arguably might as well just wear the meson scanners/engineering goggles/NVG mesons. It's too expensive for the benefit it gives under the current setup. It's nowhere near the same as thermal vision, which was made expensive intentionally due to how OP it was.
> 
> Also, as for the meson eyes - are these eyes subject to EMPs? _Wouldn't that be an extremely bad idea for the chief engineer if they, for example wanted to work with a singularity, which emits EMPs often_?

I think it does warrant some expense, like around 25 - 30 since you're still seeing the layout while benefiting from other valuable upgrades. Any knowledge beyond what you see is extremely powerful and it does show things like cult floors too, but I can agree 40 might be a little excessive.

As for the CE, what about giving their locker a pair of "hardened" implants in an auto-implanter?

**Comment by KrabSpider** [comment]
The meson vision just sounds rather pointless, especially with the price tacked onto it + the recipe involved. If this was night vision I would understand but it's just mesons...

**Comment by WanderingFox95** [comment]
> 
> 
> I'm unsure if genetics meson vision warrants 40 instability, genetics-wise - nor does it need to be locked behind a mutation of near-sighted+stimmed, compared to thermal vision. It doesn't give away the location of antagonists at all aside from the aforementioned false walls, and meson scanners on their own are very much prevalent even as just possible maintenance loot. You arguably might as well just wear the meson scanners/engineering goggles/NVG mesons. It's too expensive for the benefit it gives under the current setup. It's nowhere near the same as thermal vision, which was made expensive intentionally due to how OP it was.

Thermals weren't nearly as expensive as they should be for what they give. You can stack them with Space adaptation and Telekinesis without any repercussions, they're way too easy to get, as is anything in genetics. With enough luck, you get the best stuff before a scientist grabs the RPD from Aux Base Construction to start doing toxins and needless to say it's a huge point that has been constantly argued before. I can agree with Mesons not being worth 40 instability, but I can't agree with Thermals being as expensive as Mind Reading, something extremely useless outside of very specific situations.

> 
> Also, as for the meson eyes - are these eyes subject to EMPs? _Wouldn't that be an extremely bad idea for the chief engineer if they, for example wanted to work with a singularity, which emits EMPs often_?

What if an IPC or Synthetic wants to be CE? People know the risks, they're not stupid. You also don't really move that close to a singularity after setting it up, under normal conditions.

> 
> 
> The meson vision just sounds rather pointless, especially with the price tacked onto it + the recipe involved. If this was night vision I would understand but it's just mesons...

You know, that's... actually a good idea, let me look into it, changing the genetics side of things into that, it offers something new.

> 
> As for the CE, what about giving their locker a pair of "hardened" implants in an auto-implanter?

Making this specific implant resistant to EMPs specifically is also not a bad solution.

**Comment by BlueWildrose** [comment]
> Thermals weren't nearly as expensive as they should be for what they give. You can stack them with Space adaptation and Telekinesis without any repercussions, they're way too easy to get, as is anything in genetics. With enough luck, you get the best stuff before a scientist grabs the RPD from Aux Base Construction to start doing toxins and needless to say it's a huge point that has been constantly argued before. I can agree with Mesons not being worth 40 instability, but I can't agree with Thermals being as expensive as Mind Reading, something extremely useless outside of very specific situations.

I mean, my opinion of genetics thermals doesn't really matter in the end since it's getting removed outside of admin spawning. Usually admin-spawned genes shouldn't have any instability at all in that case - just take a look at the absurdly OP laser eyes genes and its lack of instability.

But otherwise, as my argument stands for mesons - I could just crawl maintenance and get lucky finding a pair of meson goggles there, or if I really wanted meson vision, I'd grab them from the engineering or cargo lathe or steal them from a mining closet. With how easily obtainable mesons are as an item, it shouldn't take that much effort in its current state to get them as a gene. They are permanent mesons in a sense however, so having 25 instability (what thermals originally were long ago) seems reasonable enough. To not have them be locked behind a recipe is also good.

Night vision seems like an interesting idea. We even have night-vision based spells (from revenants and whatnot) prepared you can recycle to make such a gene. Some genetics like geladikinesis, cryokinesis, and telepathy are technically spells, code-wise, so it'll be fine that way.

**Comment by WanderingFox95** [comment]
> 
> Night vision seems like an interesting idea. We even have night-vision based spells (from revenants and whatnot) prepared you can recycle to make such a gene. Some genetics like geladikinesis, cryokinesis, and telepathy are technically spells, code-wise, so it'll be fine that way.

Well, that's the next "problem".
Here are some numbers: 
Normal Nightvision as a roundstart quirk: Range: 3, lighting plane alpha 223.
Nightvision Goggles: Range: 8, lighting plane alpha: 128, flash protection: -2
Nightvision Eyes (Species like shadows, xenos or ashlizards): Range: 8, Toggle between 0, 128, 192 and 255 lighting plane alpha, no flash protection adjustments.(which doesn't seem to have a downside so I'm not sure why it doesn't just stay maxed)

Now, it's clear this can't be only like roundstart nightvision, but making it like nightvision eyes seems broken too.
Would making it like nightvision goggles be adequate? Or should it be slightly worse due to it not requiring any resources to mass-produce?

**Comment by BlueWildrose** [comment]
> Well, that's the next "problem".
> Here are some numbers:
> Normal Nightvision as a roundstart quirk: Range: 3, lighting plane alpha 223.
> Nightvision Goggles: Range: 8, lighting plane alpha: 128, flash protection: -2
> Nightvision Eyes (Species like shadows, xenos or ashlizards): Range: 8, Toggle between 0, 128, 192 and 255 lighting plane alpha, no flash protection adjustments.(which doesn't seem to have a downside so I'm not sure why it doesn't just stay maxed)
> 
> Now, it's clear this can't be only like roundstart nightvision, but making it like nightvision eyes seems broken too.
> Would making it like nightvision goggles be adequate? Or should it be slightly worse due to it not requiring any resources to mass-produce?

It's a difficult factor because yes, you get permanent night vision that way, although if it's going to replace what thermals/mesons costed, it's probably reasonable.

Being weaker to flashes as a gene however means being permanently weaker and would likely become a major inconvenience if the user ever needs to actually weld something - and yes, flash code is based on that. Did you know that if you have an experimental welder, you won't be hurt by it if you're wearing only sunglasses and not welders? If that was added, that'd be the first instance of positive genes having a downside. Maybe making use of the rarely used chromosomes so that they won't have that flash downside could be one option. Then again, this would not be available if night vision was made through a mutation recipe...

The other option I could see is letting the user toggle their night vision on and off, and thus letting them not have their flash vulnerability as a result. This could create a scenario where they can risk getting flashed by security/cyborgs in the darkness when they have night vision on, or they can choose to turn it off for that scenario perhaps, allowing night vision users to choose when they want to have that flash vulnerability or not.

I will add, the whole reason the night vision goggles was nerfed that way was because of how easily accessible it was to literally everyone, a variant of each being available in each department's protolathes with the exception of the service lathe. If you were given no downsides to wearing night vision, of course you would pick that over the default HUDs. Genetics may not be as easily accessible, especially if the geneticist/skeleton crew scientist/research director/chief medical officer/captain's judgement is being very stingy with their genes.

**Comment by WanderingFox95** [comment]
> 
> The other option I could see is letting the user toggle their night vision on and off, and thus letting them not have their flash vulnerability as a result. This could create a scenario where they can risk getting flashed by security/cyborgs in the darkness when they have night vision on, or they can choose to turn it off for that scenario perhaps, allowing night vision users to choose when they want to have that flash vulnerability or not.

Suggestion: Something like a Changeling's Augmented Eyesight that switches between nightvision and -2 flash resistance and normal vision with no downsides, for 40 instability?

**Comment by BlueWildrose** [comment]
> > The other option I could see is letting the user toggle their night vision on and off, and thus letting them not have their flash vulnerability as a result. This could create a scenario where they can risk getting flashed by security/cyborgs in the darkness when they have night vision on, or they can choose to turn it off for that scenario perhaps, allowing night vision users to choose when they want to have that flash vulnerability or not.
> 
> Suggestion: Something like a Changeling's Augmented Eyesight that switches between nightvision and -2 flash resistance and normal vision with no downsides, for 40 instability?

Sounds good enough. It'd be comparable to toggling a welding mask with a click of an action button. Or you know, taking off your NVG of choice with a click of getting them off your eyes.

**Comment by WanderingFox95** [comment]
> 
> Sounds good enough. It'd be comparable to toggling a welding mask with a click of an action button. Or you know, taking off your NVG of choice with a click of getting them off your eyes.

And done. I moved it into the action section of mutations and nabbed the augmented ling vision as a template. 
May need balancing down the line but for now it's essentially inherent, toggleable nightvision for 40 instability.

**Comment by ghost** [comment]
While nightvision is acceptable as a substitute, it should absolutely have downsides. The fact that thermal lacked it was a massive undersight, considering even the tot's thermal scanner eyeslot item is a -1 to flash protection. If NV becomes a gene, it needs a clear downside. Something like being able to toggle between NV/No NV once every 30 seconds to give a clear window for counterplay to something you can't even see or know they have

Compare the gene, to the eyeslot item which has a giant neon sign that flashes when it's worn


> But otherwise, as my argument stands for mesons - I could just crawl maintenance and get lucky finding a pair of meson goggles there, or if I really wanted meson vision, I'd grab them from the engineering or cargo lathe or steal them from a mining closet. With how easily obtainable mesons are as an item, it shouldn't take that much effort in its current state to get them as a gene. They are permanent mesons in a sense however, so having 25 instability (what thermals originally were long ago) seems reasonable enough. To not have them be locked behind a recipe is also good.

You're forgetting the key difference about genes and goggles. Genes don't take the eye slot, which means you're free to take mesons and sunglasses, or mesons and a medhud or mesons and a sechud. You're getting so much information for free while benefiting from a free eye slot

**Comment by WanderingFox95** [comment]
> 
> 
> While nightvision is acceptable as a substitute, it should absolutely have downsides. The fact that thermal lacked it was a massive undersight, considering even the tot's thermal scanner eyeslot item is a -1 to flash protection. If NV becomes a gene, it needs a clear downside. Something like being able to toggle between NV/No NV once every 30 seconds to give a clear window for counterplay to something you can't even see or know they have

It's currently at 10 seconds but adjusting that is easy to do in both directions.

> 
> Compare the gene, to the eyeslot item which has a giant neon sign that flashes when it's worn
> 

To be fair, if you're wearing a mask, you can no longer see that either, i.e. even the glass gas mask completely hides your eye equip slot, minus those that affect your eye colour (mainly scanner huds).

> > But otherwise, as my argument stands for mesons - I could just crawl maintenance and get lucky finding a pair of meson goggles there, or if I really wanted meson vision, I'd grab them from the engineering or cargo lathe or steal them from a mining closet. With how easily obtainable mesons are as an item, it shouldn't take that much effort in its current state to get them as a gene. They are permanent mesons in a sense however, so having 25 instability (what thermals originally were long ago) seems reasonable enough. To not have them be locked behind a recipe is also good.
> 
> You're forgetting the key difference about genes and goggles. Genes don't take the eye slot, which means you're free to take mesons and sunglasses, or mesons and a medhud or mesons and a sechud. You're getting so much information for free while benefiting from a free eye slot

This is why I decided to give it a cooldown, to begin with. Initially, I was confused why it had one (cause the default is 10 seconds), but then I kept it.

**Comment by BlueWildrose** [comment]
> it does show things like cult floors too

I know I'm like 6 days late but I'd like to add, this actually isn't 100% true. There's code explicitly to hide cult floors from meson vision, replacing it with a visual of plating. I know this because I recently created a dungeon for the crew to explore that used cult floors at one point, and the debug glasses with x-ray turned off (when that's turned off, it's in mesons mode) explicitly hid that. That likely means the only thing mesons can really show is where false walls would be in terms of secrecy.

Sure, mesons can show unusual activity like the layout of the walls and floors being different than usual, but speaking from experience, it outright prevents cult floors from being seen. I actually remember one time as an engineering cyborg where I discovered a cult hideout not by using mesons, but through the naked eye of just seeing a red light poking through a wall in maintenance. Cult pylons glow red.

**Comment by ghost** [comment]
> 
> 
> > it does show things like cult floors too
> 
> I know I'm like 6 days late but I'd like to add, this actually isn't 100% true. There's code explicitly to hide cult floors from meson vision, replacing it with a visual of plating. I know this because I recently created a dungeon for the crew to explore that used cult floors at one point, and the debug glasses with x-ray turned off (when that's turned off, it's in mesons mode) explicitly hid that. That likely means the only thing mesons can really show is where false walls would be in terms of secrecy.
> 
> Sure, mesons can show unusual activity like the layout of the walls and floors being different than usual, but speaking from experience, it outright prevents cult floors from being seen. I actually remember one time as an engineering cyborg where I discovered a cult hideout not by using mesons, but through the naked eye of just seeing a red light poking through a wall in maintenance. Cult pylons glow red.

How long has this been a thing for? Because I remember when we used to actually have blood rounds, that I'd see cult floors with mesons, as did others get called out because someone built a pylon causing floors to be visible when an engie or borg happened to walk by with their mesons.

**Comment by BlueWildrose** [comment]
No idea, really.

As it seems, this is how cult floors and walls work - see this webm.
https://lambda.sx/7rJ.webm

Pretty much everything is hidden except bloodcult walls. Dunno why bloodcult walls are like that really. But pylons don't convert walls at the very least.

**Comment by WanderingFox95** [comment]
> Pretty much everything is hidden except bloodcult walls. Dunno why bloodcult walls are like that really. But pylons don't convert walls at the very least.

I mean, I know from map-editing that the sprite used by clock cult walls in the editor is identical to the default wall sprite and that the special look is only added ingame. Might be some sort of turf/visible wall discrepancy?

Anyway, things are adjusted, I'm kinda just waiting till everyone's happy with them, I guess.
Mesons are a normal mutation now, Nightvision is a toggle that replaces the thermals recipe.

**Comment by YakumoChen** [comment]
Disagree with this heavily mainly because there have already been several PRs in this department that have tackled the issue and have been merged / denied, based on precedence I don't really think this PR needs to nerf it further.

#14430
#13818

Thermals is no longer a "free" mutation, X-ray is already the illegal implant version of thermals, which is a combat implant. 
I don't really believe mesons are a "combat" implant, either. Mesons are freely available to engineers already and several other jobs

TG does have a rebalance of thermals you should look at, it does eye damage and only lets you have it for a few seconds with a long cooldown.

**Comment by ghost** [comment]
I do hope we get this back in. It was nice not being able to obtain something within the first 10 minutes of the round consistently that invalidates any notion of espionage and stealth

**Comment by WanderingFox95** [comment]
Testmerge again please.

File: code/datums/mutations/actions.dm [hunk 1]
```diff
@@ -502,3 +502,43 @@
...
+	active = FALSE
+	action_icon = 'icons/mob/actions/actions_changeling.dmi'
+	action_icon_state = "ling_augmented_eyesight"
+
+/obj/effect/proc_holder/spell/self/nvtoggle/cast(mob/living/carbon/human/user)
+	. = ..()
+	var/obj/item/organ/eyes/E = user.getorganslot(ORGAN_SLOT_EYES)
+	if(E)
+		if(!active)
+			ADD_TRAIT(user, TRAIT_TRUE_NIGHT_VISION, GENETIC_MUTATION)
```
Comment (DeltaFire15) [inline code/datums/mutations/actions.dm:529]: Doesn't this mean you can activate this, then remove your eyes and replace them with normal ones to gain night vision without any negatives? Should probably adjust the eyes' see_in_dark instead, plus react to them being removed.

File: code/datums/mutations/sight.dm [hunk 2]
```diff
@@ -42,7 +41,6 @@
 	difficulty = 18
 	text_gain_indication = "<span class='notice'>You can see the heat rising off of your skin...</span>"
 	time_coeff = 2
-	instability = 40
```
Comment (DeltaFire15) [inline code/datums/mutations/sight.dm]: Leave instability the same as they're still the same effectiveness use-wise if someome *somehow* gains these (and if admins don't want it to have it as exception during bussing they can just manually adjust it)

File: code/datums/mutations/sight.dm [hunk 3]
```diff
@@ -59,12 +57,22 @@
...
+	instability = 25
+	locked = FALSE
+	visionflag = TRAIT_MESON_VISION
+
 ///X-ray Vision lets you see through walls.
 /datum/mutation/human/thermal/x_ray
 	name = "X Ray Vision"
 	desc = "A strange genome that allows the user to see between the spaces of walls." //actual x-ray would mean you'd constantly be blasting rads, wich might be fun for later //hmb
 	text_gain_indication = "<span class='notice'>The walls suddenly disappear!</span>"
-	instability = 50
```
Comment (DeltaFire15) [inline code/datums/mutations/sight.dm:67]: Same here.

**Review by DeltaFire15 — CHANGES_REQUESTED** [review]
After considerable deliberation, this seems alright. Some changes will need to be made though to avoid wackyness, afterwards testmerging should be fine.

File: code/datums/mutations/actions.dm (hunk 1 above)
Comment (WanderingFox95) [reply code/datums/mutations/actions.dm:529]: This was already very confusing to do when I did it 6 months ago and I have no idea how to adjust all that much about it anymore.

Comment (DeltaFire15) [reply code/datums/mutations/actions.dm:529]: I can take a look soonish:tm: and see what I can do, ping me tomorrow to remind me if I'm not done by then.

Comment (Linzolle) [reply code/datums/mutations/actions.dm:529]: Any word on this? Changing see_in_dark would be very easy

**Comment by keronshb** [comment]
One year anniversary

**Comment by TheEaterOfBeans** [comment]
old PR
old PR

**Comment by WanderingFox95** [comment]
Incredibly old and outdated and I don't know how to fix it, plus it's not been synced for over a year. But hey, I tried.

**Pull request closed without merging**